File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(height(root)==-1)
        {
            return false;
        }
        else
        {
            return true;
        }   
    }

    int height(TreeNode*root)
    {
        if(root==NULL)
        {
            return 0;
        }
        int left=height(root->left);
        if(left==-1)
        {
            return -1;
        }
        int right=height(root->right);
        if(right==-1)
        {
            return -1;
        }
        int ans=abs(left-right);
        if(ans>1)
        {
            return -1;
        }
        else
        {
            return max(left,right)+1;
        }
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer>ans=new ArrayList<>();
        for(int x:nums){
            int left=0;
            int right=ans.size();

            while(left<right){
                int mid=left+(right-left)/2;
                if(ans.get(mid)>=x){
                    right=mid;
                }
                else{
                    left=mid+1;
                }
            }
            if(left==ans.size()){
                ans.add(x);
            }
            else{
                ans.set(left,x);
            }
        }
        return ans.size();
    }
    
}
